cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/indexer/ParentIndexer.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright © 2018 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.data2.metadata.indexer;
+
+import co.cask.cdap.api.metadata.MetadataEntity;
+import co.cask.cdap.data2.metadata.dataset.MetadataDataset;
+import co.cask.cdap.data2.metadata.dataset.MetadataEntry;
+import co.cask.cdap.data2.metadata.dataset.SortInfo;
+import co.cask.cdap.proto.id.EntityId;
+
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * Indexer used to index a {@link MetadataEntity} with it's parent. The parent is determined by
+ * {@link EntityId#getSelfOrParentEntityId(MetadataEntity)}.
+ */
+public class ParentIndexer implements Indexer {
+
+  public static final String PARENT_KEY = "parent_entity";
+
+  @Override
+  public Set<String> getIndexes(MetadataEntry entry) {
+    Set<String> indexes = new HashSet<>();
+    EntityId selfOrParentEntityId = EntityId.getSelfOrParentEntityId(entry.getMetadataEntity());
```

Comment:
I think you only need to index this for custom entities? If the entity itself is a known entity, then you would otherwise index itself as its own parent?

Reply:
Also, this is not truly the parent. It is the nearest ancestor of known type. Maybe call it "know_entity" instead of "parent_entity"?

Ona separate note, we would need the parent entity if we want to support meta data browsing, where you can list all direct sub-entities (children) of any entity. For that we would have to index the parent of each entity, so we should not use "parent_entity" now.

Reply:
Just to clarify, this will create an index row that starts with ```parent_entity```, that is, all index entries of this kind will be in a contiguous range of the index?

Reply:
Is this indexer applied to the user metadata or the system metadata? Or both?

Reply:
> Is this indexer applied to the user metadata or the system metadata? Or both?

It is applied to both.

Reply:
> Just to clarify, this will create an index row that starts with parent_entity, that is, all index entries of this kind will be in a contiguous range of the index?

Yes.

Reply:
> Also, this is not truly the parent. It is the nearest ancestor of known type. Maybe call it "know_entity" instead of "parent_entity"?
Ona separate note, we would need the parent entity if we want to support meta data browsing, where you can list all direct sub-entities (children) of any entity. For that we would have to index the parent of each entity, so we should not use "parent_entity" now.

Done.

Reply:
> I think you only need to index this for custom entities? If the entity itself is a known entity, then you would otherwise index itself as its own parent?

Changed to index only custom entities.